### PR TITLE
Add support for TopN to cost

### PR DESCRIPTION
**Pull request by kabunchi**

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import com.google.common.collect.ImmutableList;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.plan.TopNNode;
+import org.testng.annotations.Test;
+
+public class TestTopNStatsRule
+        extends BaseStatsCalculatorTest
+{
+    @Test
+    public void testShorterTopN()
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java:26]: Name `testTopNWithLongInput`

File: presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                        .symbolStats("i2", assertion -> assertion
+                                .lowValue(0)
+                                .highValue(3)
+                                .dataSizeUnknown()
+                                .distinctValuesCount(4)
+                                .nullsFraction(0)));
+    }
+
+    @Test
+    public void testLongerTopN()
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java:96]: Name `testTopNWithSmallInput`

File: presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: I think you should reset nulls fraction to 0 here. Or is it done anyway somewhere else. Can you please add test where input has non-zero null fraction.

File: presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java (hunk 1 above)
Comment (kabunchi) [reply presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java:26]: fixed

File: presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java (hunk 2 above)
Comment (kabunchi) [reply presto-main/src/test/java/io/prestosql/cost/TestTopNStatsRule.java:96]: fixed

File: presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java (hunk 3 above)
Comment (kabunchi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: not sure I follow.
I completely disregard the null fraction as I validate only against the outputRows. adding it to a test won't really affect the changes.
I can change one of the stats in the test to hold fraction other than 0..
btw, I just followed LimitStatsRule.java

Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: Reseting nulls fraction to 0 is an overstatement. But IMO nulls handling could be improved here.

Let use two example queries:
```
presto:default> select * from (values 1, NULL, 2, NULL, 3, NULL, 4, NULL, 5) t(a) order by a limit 6;
  a
------
    1
    2
    3
    4
    5
 NULL
```

```
presto:default> select * from (values 1, NULL, 2, NULL, 3, NULL, 4, NULL, 5) t(a) order by a desc limit 6;
  a
------
    5
    4
    3
    2
    1
 NULL
(6 rows)
```

Here we have a `TopN` operator on column `a`. No matter what sort order we are using, `NULLs` will always end up last. So estimated nulls-fraction for an output should be adjusted. Something like this should do:
```
if limit > input.rowsCount {
  return input stats
}

inputNonNullsCount = input.rowCount - input.rowCount * input.nullsFraction;;

outputRowCount = limit
if limit < inputNonNullsCount {
   outputNullsCount = 0
} else {
   outputNullsCount = limit - inputNonNullsCount
}

outputNullsFraction = outputNullsCount/outputRowCount
```

Note that if ordering uses multiple columns, above logic is valid only for the first one.

Comment (kabunchi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: Got you now.
Still now sure how to proceed.
Mainly because I didn't fully understood the first column comment, how do I differ? what the single null fraction value stands for in case of multiple columns?
Regardless,  because one can also use `order by 1 NULLS first` I must look at the sortOrder to re-calculate the null fraction correctly

Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: > Mainly because I didn't fully understood the first column comment, how do I differ? what the single null fraction value stands for in case of multiple columns?

There is no single nulls-fraction value. The nulls-fraction is stored per column. You can modify it either using `PlanNodeStatsEstimate.mapSymbolColumnStatistics` or explicitly setting via `PlanNodeStatsEstimate.Builder#addSymbolStatistics()`.
If there are multiple columns used for ordering, I think we cannot derive anything reasonable regarding nulls fraction for any column but first. So I think we should leave source null-fraction for other ones.

> Regardless, because one can also use `order by 1 NULLS first` I must look at the sortOrder to re-calculate the null fraction correctly

Good point.

Comment (kabunchi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: ended up a bit ugly:
 ```
Symbol symbol = node.getOrderingScheme().getOrderBy().get(0); // Assuming not empty list
        SymbolStatsEstimate symbolStatsEstimate = sourceStats.getSymbolStatistics().get(symbol);
        SortOrder sortOrder = node.getOrderingScheme().getOrdering(symbol);
        long count = node.getCount();

        if ((symbolStatsEstimate == null) || (sortOrder == null)) {
            return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
                                                    .setOutputRowCount(count)
                                                    .build());
        }

        double nullCount = rowCount * symbolStatsEstimate.getNullsFraction();
        double outNullFraction;

        if (sortOrder.isNullsFirst()) {
            if (nullCount > count) {
                outNullFraction = 1;
            }
            else {
                outNullFraction = nullCount / count;
            }
        }
        else {
            double nonNullCount = (rowCount - nullCount);
            if (nonNullCount > count) {
                outNullFraction = 0;
            }
            else {
                outNullFraction = (count - nonNullCount) / count;
            }
        }
        SymbolStatsEstimate newSymbolStatsEstimate = SymbolStatsEstimate.buildFrom(symbolStatsEstimate).setNullsFraction(outNullFraction).build();

        // TopN actually limits (or when there was no row count estimated for source)
        return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
                .setOutputRowCount(count)
                .addSymbolStatistics(symbol, newSymbolStatsEstimate)
                .build());
    }

```
still think it worth it?

Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:45]: You can make it (IMO) a bit cleaner with `mapSymbolColumStatistics`. Like this:
```java
    @Override
    protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
    {
        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
        if (sourceStats.getOutputRowCount() <= node.getCount()) {
            return Optional.of(sourceStats);
        }

        Symbol firstOrderSymbol = node.getOrderingScheme().getOrderBy().get(0); // Assuming not empty list
        SortOrder sortOrder = node.getOrderingScheme().getOrdering(firstOrderSymbol);
        long limitCount = node.getCount();

        PlanNodeStatsEstimate resultStats = PlanNodeStatsEstimate.buildFrom(sourceStats)
                .setOutputRowCount(limitCount)
                .build();

        // augment null fraction estimation for first ORDER BY symbol
        double rowCount = sourceStats.getOutputRowCount();
        resultStats = resultStats.mapSymbolColumnStatistics(firstOrderSymbol, symbolStats -> {
            SymbolStatsEstimate.Builder newStats = SymbolStatsEstimate.buildFrom(symbolStats);
            double nullCount = rowCount * symbolStats.getNullsFraction();
            if (sortOrder.isNullsFirst()) {
                if (nullCount > limitCount) {
                    newStats.setNullsFraction(1.0);
                }
                else {
                    newStats.setNullsFraction(nullCount / limitCount);
                }
            }
            else {
                double nonNullCount = (rowCount - nullCount);
                if (nonNullCount > limitCount) {
                    newStats.setNullsFraction(0.0);
                }
                else {
                    newStats.setNullsFraction((limitCount - nonNullCount) / limitCount);
                }
            }
            return newStats.build();
        });
        return Optional.of(resultStats);
    }
```

I think it is nice to have. What needs to be added to above is handling division by 0.0. And testing.

Comment (kabunchi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:47]: fixed and pushed.
thanks for the very responsive review.

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 4]
```diff
@@ -259,13 +259,17 @@ public TopNNode topN(long count, List<Symbol> orderBy, PlanNode source)
 
     public TopNNode topN(long count, List<Symbol> orderBy, TopNNode.Step step, PlanNode source)
     {
-        return new TopNNode(
+        return topN(count, orderBy, step, source, SortOrder.ASC_NULLS_FIRST);
+    }
+
+    public TopNNode topN(long count, List<Symbol> orderBy, TopNNode.Step step, PlanNode source, SortOrder sortOrder)
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java:265]:
```suggestion
    public TopNNode topN(long count, List<Symbol> orderBy, TopNNode.Step step, SortOrder sortOrder, PlanNode source)
```

File: presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        double rowCount = sourceStats.getOutputRowCount();
+
+        if ((rowCount <= node.getCount()) || (node.getStep() == TopNNode.Step.PARTIAL)) {
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: Not sure about `PARTIAL` handling. It does not limit output size to same extent as `FINAL`, yet it still can very substantially limit it vs size of the input to `queryParalellism*limit`. @findepi do you have an idea how to do better estimation here?

**Review by losipiuk — COMMENTED** [review]
Thanks. It looks mostly good to me. @findepi can you take a look too?

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 4 above)
Comment (kabunchi) [reply presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java]: fixed and pushed.

File: presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java (hunk 5 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: There's no good answer for PARTIAL. The number of global rows coming out of that operation will depend on the number of tasks that get created at runtime, and that's not known in advance.

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: See also https://github.com/prestosql/presto/blob/c74b775e103a7a239aafef1b7ca15fe6370495fe/presto-main/src/main/java/io/prestosql/cost/AggregationStatsRule.java#L56-L58

Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: After discussing with @findepi  i think we should go same approach here:
```java
if (node.getStep() != SINGLE) {
    return Optional.empty();
}
```

It should still help as join enumeration, which uses stats is happening before `SINGLE` `TopNNodes` are split into `PARTIAL` and `FINAL` (unless I am misreading the code). (*)


(*) It looks like the only way `TopNNode` can be converted into `PARTIAL`, before join enumeration, is due to execution of `PushLimitThroughOuterJoin` or `PushLimitThroughUnion`, which enforce `partial` flag on `LimitNode` they produce. Not sure if that logic is needed - I will look more into that.

Comment (kabunchi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: just wondering why `Optional.empty()` and not `Optional.of(sourceStats);`?

Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: `Optional.empty()` means we cannot compute statistics which is accurate. The `Optional.of(sourceStats)` is not correct as the output of `PARTIAL` aggregation does not match its input.

Comment (kabunchi) [reply presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java:50]: got it. thanks.

**Comment by kabunchi** [comment]
anything else needed from my end?

**Comment by losipiuk** [comment]
> anything else needed from my end?

Oh :) Can you please make this change so we are only estimating statistics for aggregation of type `SINGLE`?

**Comment by kabunchi** [comment]
pushed. (thought I already did, was lying in my local branch) thanks.

**Comment by losipiuk** [comment]
Merged, thanks!
